apollo-router/src/spec/schema.rs

```diff
@@ -38,24 +42,353 @@ pub(crate) struct Schema {
+            // TODO: error if not found?
+            if let Some(join_enum) = compiler.db.find_enum_by_name("join__Graph".into()) {
+                for (name, url) in join_enum
+                    .enum_values_definition()
+                    .iter()
+                    .filter_map(|value| {
+                        let join_directive = value
+                            .directives()
+                            .iter()
+                            .find(|directive| directive.name() == "join__graph")?;
+                        let name = as_string(join_directive.argument_by_name("name")?)?;
+                        let url = as_string(join_directive.argument_by_name("url")?)?;
+                        Some((name, url))
+                    })
+                {
+                    if url.is_empty() {
+                        return Err(SchemaError::MissingSubgraphUrl(name.clone()));
+                    }
+                    let url = Uri::from_str(url)
+                        .map_err(|err| SchemaError::UrlParse(name.clone(), err))?;
+                    if subgraphs.insert(name.clone(), url).is_some() {
+                        return Err(SchemaError::Api(format!(
+                            "must not have several subgraphs with same name '{}'",
+                            name
+                        )));
+                    }
+                }
+            }
+
+            let object_field =
+                |field: &hir::FieldDefinition| (field.name().to_owned(), field.ty().into());
+            let implements_interface = |imp: &hir::ImplementsInterface| imp.interface().to_owned();
+            let object_types: HashMap<_, _> = compiler
+                .db
+                .object_types()
+                .iter()
+                .map(|(name, def)| {
+                    (
+                        name.clone(),
+                        ObjectType {
+                            name: name.clone(),
+                            fields: def
+                                .fields_definition()
+                                .iter()
+                                .chain(
+                                    def.extensions()
+                                        .iter()
+                                        .flat_map(|ext| ext.fields_definition()),
+                                )
+                                .map(object_field)
+                                .collect(),
+                            interfaces: def
+                                .implements_interfaces()
+                                .iter()
+                                .chain(
+                                    def.extensions()
+                                        .iter()
+                                        .flat_map(|ext| ext.implements_interfaces()),
+                                )
+                                .map(implements_interface)
+                                .collect(),
```

Comment:
I'm wondering if it would be more readable if we used `From` impl for that kind of code ?

Reply:
Done! Indeed it looks better